Optimize increment summations [Latest Nov 15]

laithsakka opened the pull request:
Summary:
**wins**
on torchrec benchmark, for 2K nodes it save 40seconds
with the recent sympy changes (https://www.internalfb.com/diff/D65883538) we save around 13 second ( with the max opt on). 
```
buck2 run fbcode//mode/opt fbcode//torchrec/distributed/tests:pt2_compile_benchmark -- --num-features=200
```
This diff optimizes construction expressions of the form 
a+b+c...  (all unique symbols). 
which are very common in torchrec models. 

**How**
Expressions of the form a+b+c are not optimized by add, the only needed optimization is sorting them.
If we have  a+b+c and we are adding (d) to it, we can do a binary search to know 
the position of (d) and avoid optimizing the new expression by passing the new order. 


**Extensions**:
1. support constant terms.
2. support 10a+10b+.. (this will give even more wins will extend the support in second PR)

Differential Revision: D66008482




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ezyang @SherlockNoMad @EikanWang @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

laithsakka commented:
offline discussion todos:
1. document why we are using  getattr(self, "_optimized_summation", False), on the symNode.
2. add micro benchmark.

ezyang commented on torch/fx/experimental/sym_node.py:
```diff
@@ -1134,6 +1202,13 @@ def binary_magic_impl(self, other):
+                    self.expr,
+                    other.expr,
+                    getattr(self, "_optimized_summation", False),
+                    getattr(other, "_optimized_summation", False),
```
We discussed this in person, where you defended the ad hoc getattr/setattr because it was only set and accessed from two places. I think my preferred way of documenting this sort of situation is to have a `# Note [blah blah blah]` in one location describing the situation (probably the comment below), and then referencing that note consistently at both use sites.

The important thing to document, which is not directly documented at either of these sites, is what exactly the invariant specified by optimized summation is.

ezyang replied:
Yeah, I think I would even be happy with "this is a subclass of Add and is identical to Add in all respects except it respects the optimized summation invariant". This would probably have some annoying side effects in other parts of the code but from a layering perspective it's much cleaner.

laithsakka replied:
i tried the subclass it did not work, because we could get an add to co-live with custom add and they dont compare to be equal

ezyang commented on torch/fx/experimental/sym_node.py:
```diff
@@ -755,8 +755,75 @@ def _sympy_rshift(a, b):
+    is when we know the expression is a summation of unique symbols all we need to know is the correct order of symbols,
+    no other optimizations is needed. Figuring out the order can be done in Log(n) comparisons when its known that the input
+    is already sorted. Returns a tuple of (1) a boolean that indicates whether the output is a summation of unique symbols,
+    (2) the result sympy expression.
```
Specifically you want to talk about how comparisons on sympy expressions sortkey are expensive

ezyang commented on torch/fx/experimental/sym_node.py:
```diff
@@ -755,8 +755,75 @@ def _sympy_rshift(a, b):
+            rhs._args[0]
+        ):
+            # (a0+a1) + (a2+a3) => (a0+a1+a2+a3)
+            return make_optimized(lhs._args + rhs._args)
```
Is it cheap to test the other way too? (You have a cliff here if I accidentally swap the orders of the arguments to successive add here)

laithsakka replied:
done

ezyang commented on torch/fx/experimental/sym_node.py:
```diff
@@ -755,8 +755,75 @@ def _sympy_rshift(a, b):
+    low, high = 0, len(ordered_args) - 1
+    from sympy.core.basic import _args_sortkey as sort_key
+
+    # Fast path when new_arg > ordered_args[-1], which is common.
```
good

ezyang replied:
(do you want the `new_arg < ordered_args[0]` case too?

laithsakka replied:
added

ezyang commented on torch/fx/experimental/sym_node.py:
```diff
@@ -755,8 +755,75 @@ def _sympy_rshift(a, b):
+
+    while low <= high:
+        mid = (low + high) // 2
+        if ordered_args[mid] == new_arg:
```
Hmm... so you're telling me how comparisons are expensive. Is there a way to only do one "compare" here (returning -1, 0, 1 depending on the situation)? Right now you have to do two. (And I guess you don't need `sort_key` for the equality test here because it's guaranteed to coincide with regular equality?

laithsakka replied:
use basic_compare directly

ezyang commented on torch/fx/experimental/sym_node.py:
```diff
@@ -755,8 +755,75 @@ def _sympy_rshift(a, b):
     return RShift(a, b)
 
 
+def _binary_search_insert_arg(ordered_args, new_arg):
```
assert len(ordered_args) != 0

ezyang reviewed (approved):
This looks functionally correct, see also my comments on the other PR.

laithsakka commented:
Address all comments

laithsakka committed:
Optimize increment summations [Latest Nov 15] (#140822)

Summary:

**wins**
on torchrec benchmark, for 2K nodes it save 40seconds
with the recent sympy changes (https://www.internalfb.com/diff/D65883538) we save around 13 second ( with the max opt on). 
```
buck2 run fbcode//mode/opt fbcode//torchrec/distributed/tests:pt2_compile_benchmark -- --num-features=200
```
This diff optimizes construction expressions of the form 
a+b+c...  (all unique symbols). 
which are very common in torchrec models. 

**How**
Expressions of the form a+b+c are not optimized by add, the only needed optimization is sorting them.
If we have  a+b+c and we are adding (d) to it, we can do a binary search to know 
the position of (d) and avoid optimizing the new expression by passing the new order. 


**Extensions**:
1. support constant terms.
2. support 10a+10b+.. (this will give even more wins will extend the support in second PR)

Test Plan:
add tests
add benchmark 
run tests

Reviewed By: ezyang

Differential Revision: D66008482

laithsakka commented:
@pytorchbot merge -f

laithsakka commented:
@pytorchbot merge -i
